File: backend/src/formulas/expression.py
"""Excel-free formula expressions: tokenize, parse, evaluate.

Formulas are arithmetic over *named parameters* -- never Excel cell
references. `resources/formula.md` documents the workbook originals in native
Excel, but those are reference material only: sheet names, `!` and `$` anchors
are rejected here on purpose, with a message that says so.

Nothing in this module calls `eval()` or `ast.parse()`. A fixed grammar is
tokenized, parsed into a small tuple AST, and walked against an allow-list of
functions -- the same guarded-parse shape as the SQL allow-listing in
`src/llm/agents/common/tools/freeform_query.py`.

Grammar (loosest to tightest):

    expression  := comparison
    comparison  := additive (("=" | "<>" | "<" | "<=" | ">" | ">=") additive)*
    additive    := multiplicative (("+" | "-") multiplicative)*
    multiplicative := unary (("*" | "/") unary)*
    unary       := ("-" | "+")* primary
    primary     := NUMBER | STRING | NAME | CALL | "(" expression ")"
    CALL        := NAME "(" [expression ("," expression)*] ")"
"""

from __future__ import annotations

import math
import re

from typing import Any

# Generic maths, not Excel-specific. Names are matched case-insensitively but
# are conventionally written upper-case in stored expressions.
FUNCTIONS: dict[str, tuple[int, int]] = {
    # name: (min args, max args); -1 max means variadic
    "MAX": (1, -1),
    "MIN": (1, -1),
    "ROUND": (1, 2),
    "CEILING": (1, 2),
    "IF": (2, 3),
    "AND": (1, -1),
    "OR": (1, -1),
    "NOT": (1, 1),
}

COMPARISONS = frozenset({"=", "<>", "<", "<=", ">", ">="})

_TOKEN_RE = re.compile(
    r"""
      (?P<space>\s+)
    | (?P<number>(?:\d+\.\d*|\.\d+|\d+)(?:[eE][+-]?\d+)?)
    | (?P<string>"[^"]*")
    | (?P<name>[A-Za-z_][A-Za-z0-9_]*)
    | (?P<op><>|<=|>=|[<>=+\-*/])
    | (?P<punct>[(),])
    """,
    re.VERBOSE,
)

# Characters that mean the author pasted a native Excel formula.
_EXCEL_HINTS = {
    "!": "sheet-qualified cell references",
    "$": "absolute cell anchors ($)",
    ":": "cell ranges (:)",
    ";": "argument separators (;)",
    "&": "text concatenation (&)",
    "%": "percent signs",
    "^": "exponent operators (^)",
}


class Token:
    __slots__ = ("kind", "value", "position")

    def __init__(self, kind: str, value: str, position: int) -> None:
        self.kind = kind
        self.value = value
        self.position = position

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"Token({self.kind}, {self.value!r}, {self.position})"


def tokenize(expression: str) -> list[Token]:
    """Split an expression into tokens, or raise ValueError."""
    if not isinstance(expression, str) or not expression.strip():
        raise ValueError("Expression is empty.")

    text = expression.strip()
    if text.startswith("="):
        raise ValueError(
            "Expressions must not start with '=' -- that is Excel syntax. "
            "Write the formula directly, e.g. ROUND(on_hand + open_po)."
        )

    tokens: list[Token] = []
    index = 0
    while index < len(text):
        match = _TOKEN_RE.match(text, index)
        if match is None:
            character = text[index]
            hint = _EXCEL_HINTS.get(character)
            if hint is not None:
                raise ValueError(
                    f"{hint.capitalize()} are not supported at position {index + 1}. "
                    "Formulas use named parameters only, not Excel cell references."
                )
            raise ValueError(
                f"Unexpected character {character!r} at position {index + 1}."
            )

        kind = match.lastgroup
        value = match.group()
        index = match.end()
        if kind == "space":
            continue
        tokens.append(Token(kind, value, match.start()))

    if not tokens:
        raise ValueError("Expression is empty.")
    return tokens


class _Parser:
    def __init__(self, tokens: list[Token]) -> None:
        self._tokens = tokens
        self._index = 0

    # -- token helpers -------------------------------------------------

    def _peek(self) -> Token | None:
        if self._index < len(self._tokens):
            return self._tokens[self._index]
        return None

    def _next(self) -> Token:
        token = self._peek()
        if token is None:
            raise ValueError("Expression ended unexpectedly.")
        self._index += 1
        return token

    def _accept(self, kind: str, *values: str) -> Token | None:
        token = self._peek()
        if token is None or token.kind != kind:
            return None
        if values and token.value not in values:
            return None
        self._index += 1
        return token

    def _expect(self, kind: str, value: str) -> Token:
        token = self._accept(kind, value)
        if token is None:
            found = self._peek()
            where = (
                f"position {found.position + 1}" if found else "the end of the expression"
            )
            raise ValueError(f"Expected {value!r} at {where}.")
        return token

    # -- grammar -------------------------------------------------------

    def parse(self) -> tuple:
        node = self._comparison()
        leftover = self._peek()
        if leftover is not None:
            raise ValueError(
                f"Unexpected {leftover.value!r} at position {leftover.position + 1}."
            )
        return node

    def _comparison(self) -> tuple:
        node = self._additive()
        while True:
            token = self._accept("op")
            if token is None:
                return node
            if token.value not in COMPARISONS:
                self._index -= 1
                return node
            node = ("bin", token.value, node, self._additive())

    def _additive(self) -> tuple:
        node = self._multiplicative()
        while True:
            token = self._accept("op", "+", "-")
            if token is None:
                return node
            node = ("bin", token.value, node, self._multiplicative())

    def _multiplicative(self) -> tuple:
        node = self._unary()
        while True:
            token = self._accept("op", "*", "/")
            if token is None:
                return node
            node = ("bin", token.value, node, self._unary())

    def _unary(self) -> tuple:
        token = self._accept("op", "-", "+")
        if token is None:
            return self._primary()
        operand = self._unary()
        if token.value == "+":
            return operand
        return ("neg", operand)

    def _primary(self) -> tuple:
        token = self._next()

        if token.kind == "number":
            return ("num", float(token.value))

        if token.kind == "string":
            return ("str", token.value[1:-1])

        if token.kind == "punct" and token.value == "(":
            node = self._comparison()
            self._expect("punct", ")")
            return node

        if token.kind == "name":
            if self._accept("punct", "("):
                return self._call(token)
            return ("name", token.value)

        raise ValueError(
            f"Unexpected {token.value!r} at position {token.position + 1}."
        )

    def _call(self, name_token: Token) -> tuple:
        name = name_token.value.upper()
        if name not in FUNCTIONS:
            allowed = ", ".join(sorted(FUNCTIONS))
            raise ValueError(
                f"Unknown function {name_token.value!r} at position "
                f"{name_token.position + 1}. Allowed functions: {allowed}."
            )

        arguments: list[tuple] = []
        if self._accept("punct", ")") is None:
            while True:
                arguments.append(self._comparison())
                if self._accept("punct", ","):
                    continue
                self._expect("punct", ")")
                break

        low, high = FUNCTIONS[name]
        count = len(arguments)
        if count < low or (high != -1 and count > high):
            expected = f"{low}" if low == high else (
                f"at least {low}" if high == -1 else f"{low} to {high}"
            )
            raise ValueError(
                f"{name} takes {expected} argument(s), got {count}."
            )
        return ("call", name, arguments)


def parse(expression: str) -> tuple:
    """Parse an expression into an AST, or raise ValueError."""
    return _Parser(tokenize(expression)).parse()


def referenced_names(node: tuple) -> set[str]:
    """Every parameter name the expression reads."""
    kind = node[0]
    if kind == "name":
        return {node[1]}
    if kind == "neg":
        return referenced_names(node[1])
    if kind == "bin":
        return referenced_names(node[2]) | referenced_names(node[3])
    if kind == "call":
        names: set[str] = set()
        for argument in node[2]:
            names |= referenced_names(argument)
        return names
    return set()


# -- evaluation --------------------------------------------------------


def _number(value: Any, context: str) -> float:
    if isinstance(value, bool):
        return 1.0 if value else 0.0
    if isinstance(value, (int, float)):
        return float(value)
    raise ValueError(f"{context} needs a number, got {value!r}.")


def _truthy(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)):
        return value != 0
    raise ValueError(f"Expected a condition, got the text {value!r}.")


def _compare(operator: str, left: Any, right: Any) -> bool:
    # Excel compares text case-insensitively; mirror that so "y" = "Y".
    if isinstance(left, str) or isinstance(right, str):
        if not (isinstance(left, str) and isinstance(right, str)):
            raise ValueError(
                f"Cannot compare text with a number: {left!r} {operator} {right!r}."
            )
        first: Any = left.casefold()
        second: Any = right.casefold()
        if operator not in {"=", "<>"}:
            raise ValueError(
                f"Text can only be compared with '=' or '<>', not {operator!r}."
            )
    else:
        first = _number(left, "Comparison")
        second = _number(right, "Comparison")

    if operator == "=":
        return first == second
    if operator == "<>":
        return first != second
    if operator == "<":
        return first < second
    if operator == "<=":
        return first <= second
    if operator == ">":
        return first > second
    return first >= second


def _excel_round(value: float, digits: float = 0.0) -> float:
    """ROUND with Excel's half-away-from-zero rule (Python rounds half to even)."""
    factor = 10.0 ** int(digits)
    scaled = value * factor
    rounded = math.floor(abs(scaled) + 0.5)
    return math.copysign(rounded, scaled) / factor


def _excel_ceiling(value: float, significance: float = 1.0) -> float:
    if significance == 0:
        return 0.0
    return math.ceil(value / significance) * significance


def _call(name: str, node_args: list[tuple], values: dict[str, Any]) -> Any:
    # IF short-circuits: the untaken branch must not be evaluated, so a guard
    # like IF(qty > 0, total / qty, 0) is safe.
    if name == "IF":
        condition = _truthy(_evaluate(node_args[0], values))
        if condition:
            return _evaluate(node_args[1], values)
        if len(node_args) == 3:
            return _evaluate(node_args[2], values)
        return False

    if name == "AND":
        for argument in node_args:
            if not _truthy(_evaluate(argument, values)):
                return False
        return True

    if name == "OR":
        for argument in node_args:
            if _truthy(_evaluate(argument, values)):
                return True
        return False

    if name == "NOT":
        return not _truthy(_evaluate(node_args[0], values))

    arguments = [_evaluate(argument, values) for argument in node_args]

    if name == "MAX":
        return max(_number(item, "MAX") for item in arguments)
    if name == "MIN":
        return min(_number(item, "MIN") for item in arguments)
    if name == "ROUND":
        digits = _number(arguments[1], "ROUND") if len(arguments) == 2 else 0.0
        return _excel_round(_number(arguments[0], "ROUND"), digits)
    if name == "CEILING":
        step = _number(arguments[1], "CEILING") if len(arguments) == 2 else 1.0
        return _excel_ceiling(_number(arguments[0], "CEILING"), step)

    raise ValueError(f"Unknown function {name}.")


def _evaluate(node: tuple, values: dict[str, Any]) -> Any:
    kind = node[0]

    if kind == "num":
        return node[1]
    if kind == "str":
        return node[1]

    if kind == "name":
        name = node[1]
        if name not in values:
            raise ValueError(f"No value supplied for parameter {name!r}.")
        return values[name]

    if kind == "neg":
        return -_number(_evaluate(node[1], values), "Negation")

    if kind == "call":
        return _call(node[1], node[2], values)

    operator = node[1]
    left = _evaluate(node[2], values)
    right = _evaluate(node[3], values)

    if operator in COMPARISONS:
        return _compare(operator, left, right)

    first = _number(left, f"Operator {operator!r}")
    second = _number(right, f"Operator {operator!r}")
    if operator == "+":
        return first + second
    if operator == "-":
        return first - second
    if operator == "*":
        return first * second
    if second == 0:
        raise ValueError("Division by zero.")
    return first / second


def evaluate(node: tuple, values: dict[str, Any]) -> Any:
    """Run a parsed expression against parameter values."""
    return _evaluate(node, values)


def evaluate_expression(expression: str, values: dict[str, Any]) -> Any:
    """Parse and run in one step."""
    return evaluate(parse(expression), values)

File: frontend/src/formulas/expression.js
/**
 * The formula evaluator, ported from `backend/src/formulas/expression.py`.
 *
 * WHY A SECOND COPY EXISTS
 * The What-If panel re-classifies 800 SKUs on every slider movement. Sending
 * that to `POST /api/formulas/{id}/evaluate` would be one request per SKU per
 * formula, so the evaluation happens here instead.
 *
 * What is duplicated is the *interpreter*, not what it interprets. There is no
 * threshold in this file, no state name, no lever name — nothing that could
 * disagree with the workbook about inventory, because it knows nothing about
 * inventory. The rules stay where they were: `resources/dbtemp/formula.json`,
 * one copy, read by both languages.
 *
 * That distinction is the whole argument for this file, and it is only worth
 * anything while the two evaluators agree. `expression.test.js` runs both over
 * the same worked examples and fails if they diverge.
 *
 * TEMPORARY
 * When the Inventory Risk backend builder lands and What-If moves server-side,
 * delete this module rather than leaving it beside the Python one. Two
 * evaluators are defensible while one of them is the only way to answer the
 * question; two evaluators out of habit are not.
 *
 * Grammar, loosest to tightest — identical to the Python original:
 *
 *   expression     := comparison
 *   comparison     := additive (("=" | "<>" | "<" | "<=" | ">" | ">=") additive)*
 *   additive       := multiplicative (("+" | "-") multiplicative)*
 *   multiplicative := unary (("*" | "/") unary)*
 *   unary          := ("-" | "+")* primary
 *   primary        := NUMBER | STRING | NAME | CALL | "(" expression ")"
 *   CALL           := NAME "(" [expression ("," expression)*] ")"
 */

/** name: [min args, max args]; -1 max means variadic. */
const FUNCTIONS = {
  MAX: [1, -1],
  MIN: [1, -1],
  ROUND: [1, 2],
  CEILING: [1, 2],
  IF: [2, 3],
  AND: [1, -1],
  OR: [1, -1],
  NOT: [1, 1],
};

const COMPARISONS = new Set(["=", "<>", "<", "<=", ">", ">="]);

// Sticky so it can be advanced by hand, matching Python's `re.match(text, i)`.
// Alternation order is load-bearing: `<>` must be tried before `<`.
const TOKEN_RE =
  /(?<space>\s+)|(?<number>(?:\d+\.\d*|\.\d+|\d+)(?:[eE][+-]?\d+)?)|(?<string>"[^"]*")|(?<name>[A-Za-z_][A-Za-z0-9_]*)|(?<op><>|<=|>=|[<>=+\-*/])|(?<punct>[(),])/y;

/** Characters that mean someone pasted a native Excel formula. */
const EXCEL_HINTS = {
  "!": "sheet-qualified cell references",
  $: "absolute cell anchors ($)",
  ":": "cell ranges (:)",
  ";": "argument separators (;)",
  "&": "text concatenation (&)",
  "%": "percent signs",
  "^": "exponent operators (^)",
};

export class FormulaError extends Error {}

export function tokenize(expression) {
  if (typeof expression !== "string" || !expression.trim()) {
    throw new FormulaError("Expression is empty.");
  }

  const text = expression.trim();
  if (text.startsWith("=")) {
    throw new FormulaError(
      "Expressions must not start with '=' — that is Excel syntax. " +
        "Write the formula directly, e.g. ROUND(on_hand + open_po).",
    );
  }

  const tokens = [];
  let index = 0;

  while (index < text.length) {
    TOKEN_RE.lastIndex = index;
    const match = TOKEN_RE.exec(text);

    if (match === null || match.index !== index) {
      const character = text[index];
      const hint = EXCEL_HINTS[character];
      if (hint !== undefined) {
        throw new FormulaError(
          `${hint[0].toUpperCase()}${hint.slice(1)} are not supported at ` +
            `position ${index + 1}. Formulas use named parameters only, not ` +
            "Excel cell references.",
        );
      }
      throw new FormulaError(
        `Unexpected '${character}' at position ${index + 1}.`,
      );
    }

    const kind = Object.keys(match.groups).find(
      (group) => match.groups[group] !== undefined,
    );
    const value = match[0];
    const start = index;
    index += value.length;

    if (kind !== "space") {
      tokens.push({ kind, value, position: start });
    }
  }

  if (tokens.length === 0) {
    throw new FormulaError("Expression is empty.");
  }
  return tokens;
}

class Parser {
  constructor(tokens) {
    this.tokens = tokens;
    this.index = 0;
  }

  peek() {
    return this.index < this.tokens.length ? this.tokens[this.index] : null;
  }

  next() {
    const token = this.peek();
    if (token === null) {
      throw new FormulaError("Expression ended unexpectedly.");
    }
    this.index += 1;
    return token;
  }

  accept(kind, ...values) {
    const token = this.peek();
    if (token === null || token.kind !== kind) return null;
    if (values.length && !values.includes(token.value)) return null;
    this.index += 1;
    return token;
  }

  expect(kind, value) {
    const token = this.accept(kind, value);
    if (token === null) {
      const found = this.peek();
      const where = found
        ? `position ${found.position + 1}`
        : "the end of the expression";
      throw new FormulaError(`Expected '${value}' at ${where}.`);
    }
    return token;
  }

  parse() {
    const node = this.comparison();
    const leftover = this.peek();
    if (leftover !== null) {
      throw new FormulaError(
        `Unexpected '${leftover.value}' at position ${leftover.position + 1}.`,
      );
    }
    return node;
  }

  comparison() {
    let node = this.additive();
    for (;;) {
      const token = this.accept("op");
      if (token === null) return node;
      if (!COMPARISONS.has(token.value)) {
        this.index -= 1;
        return node;
      }
      node = ["bin", token.value, node, this.additive()];
    }
  }

  additive() {
    let node = this.multiplicative();
    for (;;) {
      const token = this.accept("op", "+", "-");
      if (token === null) return node;
      node = ["bin", token.value, node, this.multiplicative()];
    }
  }

  multiplicative() {
    let node = this.unary();
    for (;;) {
      const token = this.accept("op", "*", "/");
      if (token === null) return node;
      node = ["bin", token.value, node, this.unary()];
    }
  }

  unary() {
    const token = this.accept("op", "-", "+");
    if (token === null) return this.primary();
    const operand = this.unary();
    return token.value === "+" ? operand : ["neg", operand];
  }

  primary() {
    const token = this.next();

    if (token.kind === "number") return ["num", Number(token.value)];
    if (token.kind === "string") return ["str", token.value.slice(1, -1)];

    if (token.kind === "punct" && token.value === "(") {
      const node = this.comparison();
      this.expect("punct", ")");
      return node;
    }

    if (token.kind === "name") {
      if (this.accept("punct", "(")) return this.call(token);
      return ["name", token.value];
    }

    throw new FormulaError(
      `Unexpected '${token.value}' at position ${token.position + 1}.`,
    );
  }

  call(nameToken) {
    const name = nameToken.value.toUpperCase();
    if (!(name in FUNCTIONS)) {
      const allowed = Object.keys(FUNCTIONS).sort().join(", ");
      throw new FormulaError(
        `Unknown function '${nameToken.value}' at position ` +
          `${nameToken.position + 1}. Allowed functions: ${allowed}.`,
      );
    }

    const args = [];
    if (this.accept("punct", ")") === null) {
      for (;;) {
        args.push(this.comparison());
        if (this.accept("punct", ",")) continue;
        this.expect("punct", ")");
        break;
      }
    }

    const [low, high] = FUNCTIONS[name];
    if (args.length < low || (high !== -1 && args.length > high)) {
      const expected =
        low === high
          ? `${low}`
          : high === -1
            ? `at least ${low}`
            : `${low} to ${high}`;
      throw new FormulaError(
        `${name} takes ${expected} argument(s), got ${args.length}.`,
      );
    }
    return ["call", name, args];
  }
}

/** Parse an expression into an AST, or throw FormulaError. */
export function parse(expression) {
  return new Parser(tokenize(expression)).parse();
}

/** Every parameter name the expression reads. */
export function referencedNames(node) {
  const [kind] = node;
  if (kind === "name") return new Set([node[1]]);
  if (kind === "neg") return referencedNames(node[1]);
  if (kind === "bin") {
    return new Set([...referencedNames(node[2]), ...referencedNames(node[3])]);
  }
  if (kind === "call") {
    const names = new Set();
    for (const argument of node[2]) {
      for (const name of referencedNames(argument)) names.add(name);
    }
    return names;
  }
  return new Set();
}

// -- evaluation --------------------------------------------------------

function asNumber(value, context) {
  if (typeof value === "boolean") return value ? 1 : 0;
  if (typeof value === "number" && Number.isFinite(value)) return value;
  throw new FormulaError(`${context} needs a number, got ${display(value)}.`);
}

function truthy(value) {
  if (typeof value === "boolean") return value;
  if (typeof value === "number") return value !== 0;
  throw new FormulaError(`Expected a condition, got the text ${display(value)}.`);
}

function display(value) {
  return typeof value === "string" ? `'${value}'` : String(value);
}

function compare(operator, left, right) {
  let first;
  let second;

  if (typeof left === "string" || typeof right === "string") {
    if (typeof left !== "string" || typeof right !== "string") {
      throw new FormulaError(
        `Cannot compare text with a number: ${display(left)} ${operator} ` +
          `${display(right)}.`,
      );
    }
    // Excel compares text case-insensitively; mirror that so "y" = "Y".
    first = left.toLowerCase();
    second = right.toLowerCase();
    if (operator !== "=" && operator !== "<>") {
      throw new FormulaError(
        `Text can only be compared with '=' or '<>', not '${operator}'.`,
      );
    }
  } else {
    first = asNumber(left, "Comparison");
    second = asNumber(right, "Comparison");
  }

  switch (operator) {
    case "=":
      return first === second;
    case "<>":
      return first !== second;
    case "<":
      return first < second;
    case "<=":
      return first <= second;
    case ">":
      return first > second;
    default:
      return first >= second;
  }
}

/** ROUND with Excel's half-away-from-zero rule (JS rounds half up). */
function excelRound(value, digits = 0) {
  const factor = 10 ** Math.trunc(digits);
  const scaled = value * factor;
  const rounded = Math.floor(Math.abs(scaled) + 0.5);
  return (scaled < 0 ? -rounded : rounded) / factor;
}

function excelCeiling(value, significance = 1) {
  if (significance === 0) return 0;
  return Math.ceil(value / significance) * significance;
}

function callFunction(name, nodeArgs, values) {
  // IF short-circuits: the untaken branch must not be evaluated, so a guard
  // like IF(qty > 0, total / qty, 0) is safe.
  if (name === "IF") {
    if (truthy(evaluateNode(nodeArgs[0], values))) {
      return evaluateNode(nodeArgs[1], values);
    }
    return nodeArgs.length === 3 ? evaluateNode(nodeArgs[2], values) : false;
  }

  if (name === "AND") {
    for (const argument of nodeArgs) {
      if (!truthy(evaluateNode(argument, values))) return false;
    }
    return true;
  }

  if (name === "OR") {
    for (const argument of nodeArgs) {
      if (truthy(evaluateNode(argument, values))) return true;
    }
    return false;
  }

  if (name === "NOT") return !truthy(evaluateNode(nodeArgs[0], values));

  const args = nodeArgs.map((argument) => evaluateNode(argument, values));

  if (name === "MAX") {
    return Math.max(...args.map((item) => asNumber(item, "MAX")));
  }
  if (name === "MIN") {
    return Math.min(...args.map((item) => asNumber(item, "MIN")));
  }
  if (name === "ROUND") {
    const digits = args.length === 2 ? asNumber(args[1], "ROUND") : 0;
    return excelRound(asNumber(args[0], "ROUND"), digits);
  }
  if (name === "CEILING") {
    const step = args.length === 2 ? asNumber(args[1], "CEILING") : 1;
    return excelCeiling(asNumber(args[0], "CEILING"), step);
  }

  throw new FormulaError(`Unknown function ${name}.`);
}

function evaluateNode(node, values) {
  const [kind] = node;

  if (kind === "num" || kind === "str") return node[1];

  if (kind === "name") {
    const name = node[1];
    if (!(name in values)) {
      throw new FormulaError(`No value supplied for parameter '${name}'.`);
    }
    return values[name];
  }

  if (kind === "neg") {
    return -asNumber(evaluateNode(node[1], values), "Negation");
  }

  if (kind === "call") return callFunction(node[1], node[2], values);

  const operator = node[1];
  const left = evaluateNode(node[2], values);
  const right = evaluateNode(node[3], values);

  if (COMPARISONS.has(operator)) return compare(operator, left, right);

  const first = asNumber(left, `Operator '${operator}'`);
  const second = asNumber(right, `Operator '${operator}'`);

  switch (operator) {
    case "+":
      return first + second;
    case "-":
      return first - second;
    case "*":
      return first * second;
    default:
      if (second === 0) throw new FormulaError("Division by zero.");
      return first / second;
  }
}

/** Run a parsed expression against parameter values. */
export function evaluate(node, values) {
  return evaluateNode(node, values);
}

/** Parse and run in one step. */
export function evaluateExpression(expression, values) {
  return evaluate(parse(expression), values);
}
